File: src/main/java/algo/backtrack/Subsets_78.java
package algo.backtrack;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Subsets_78 {
  public static List<List<Integer>> subsets(int[] nums) {
    var result = new ArrayList<List<Integer>>();
    Arrays.sort(nums);
    backtrack(nums, 0, result, new ArrayList<>());
    return result;
  }

  private static void backtrack(int[] nums, int start, List<List<Integer>> lists, List<Integer> list) {
    lists.add(new ArrayList<>(list));
    for (var i = start; i < nums.length; i++) {
      list.add(nums[i]);
      backtrack(nums, i + 1, lists, list);
      list.removeLast();
    }
    Arrays.sort(nums);
  }

  public static void main(String[] args) {
    subsets(new int[]{1,2}).forEach(System.out::println);
  }
}

File: src/main/kotlin/algo/backtrack/Subsets_78.kt
package algo.backtrack

class KtSubsets_78 {
    fun subsets(nums: IntArray): List<List<Int>> {
        val all: MutableList<List<Int>> = mutableListOf()
        fun backtrack(nums: IntArray, start: Int, list: MutableList<Int>) {
            all.add(list.toList())
            for (i in start..<nums.size) {
                list.add(nums[i])
                backtrack(nums, i + 1, list)
                list.removeLast()
            }
        }
        backtrack(nums, 0, mutableListOf())
        return all
    }
}

fun main() {
    val subset = KtSubsets_78()
    println(subset.subsets(intArrayOf(1,2,3)))
    println(subset.subsets(intArrayOf(0)))
}